modules/transactions/src/main/java/org/apache/ignite/internal/tx/storage/state/rocksdb/TxStateRocksDbSharedStorage.java

```diff
@@ -0,0 +1,193 @@
+            ExecutorService threadPool,
+            IntSupplier flushDelaySupplier
+    ) {
+        this.dbPath = dbPath;
+        this.scheduledExecutor = scheduledExecutor;
+        this.threadPool = threadPool;
+        this.flushDelaySupplier = flushDelaySupplier;
+
+        try {
+            Files.createDirectories(dbPath);
+        } catch (IOException e) {
+            throw new IgniteInternalException("Failed to create transaction state storage directory", e);
+        }
+    }
+
+    /**
+     * Returns shared {@link RocksDB} instance.
+     */
+    RocksDB db() {
+        return db;
+    }
+
+    /**
+     * Returns a future to await flush.
+     */
+    CompletableFuture<Void> awaitFlush(boolean schedule) {
+        return flusher.awaitFlush(schedule);
+    }
+
+    /**
+     * Starts the storage.
+     */
+    public void start() {
+        try {
+            flusher = new RocksDbFlusher(
+                    busyLock,
+                    scheduledExecutor,
+                    threadPool,
+                    flushDelaySupplier,
+                    () -> {} // No-op.
+            );
+
+            this.dbOptions = new DBOptions()
+                    .setCreateIfMissing(true)
+                    .setAtomicFlush(true)
+                    .setListeners(List.of(flusher.listener()));
+
+            List<ColumnFamilyDescriptor> cfDescriptors;
+
+            try (Options opts = new Options()) {
+                cfDescriptors = RocksDB.listColumnFamilies(opts, dbPath.toAbsolutePath().toString())
+                        .stream()
+                        .map(nameBytes -> new ColumnFamilyDescriptor(nameBytes, new ColumnFamilyOptions()))
+                        .collect(toList());
+
+                cfDescriptors = cfDescriptors.isEmpty()
+                        ? List.of(new ColumnFamilyDescriptor(TX_STATE_CF.getBytes(StandardCharsets.UTF_8), new ColumnFamilyOptions()))
+                        : cfDescriptors;
+            }
+
+            List<ColumnFamilyHandle> cfHandles = new ArrayList<>(cfDescriptors.size());
+
+            this.db = RocksDB.open(dbOptions, dbPath.toString(), cfDescriptors, cfHandles);
+
+            flusher.init(db, cfHandles);
+        } catch (Exception e) {
+            throw new IgniteInternalException("Could not create transaction state storage", e);
+        }
+    }
+
+    @Override
+    public void close() {
+        if (!stopGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+
+        try {
+            List<AutoCloseable> resources = new ArrayList<>();
+
+            resources.add(readOptions);
+            resources.add(writeOptions);
+            resources.add(dbOptions);
+            resources.add(db);
+
+            reverse(resources);
+            IgniteUtils.closeAll(resources);
+        } catch (Exception e) {
+            throw new IgniteInternalException("Failed to stop transaction state storage", e);
```

Comment:
Suggestion: Maybe just leave the `Exception` in the signature and remove this code?

Reply:
Then I would have to handle it in every invocation, and that code is not ready for checked exceptions unfortunately